caas/ecs/application.go

```diff
@@ -0,0 +1,637 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package ecs
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/ecs"
+	"github.com/aws/aws-sdk-go/service/ecs/ecsiface"
+	"github.com/juju/clock"
+	"github.com/juju/errors"
+	"github.com/kr/pretty"
+
+	"github.com/juju/juju/caas"
+	"github.com/juju/juju/core/paths"
+	"github.com/juju/juju/core/status"
+	"github.com/juju/juju/core/watcher"
+	"github.com/juju/juju/core/watcher/watchertest"
+	jujustorage "github.com/juju/juju/storage"
+)
+
+var (
+	jujuDataDir = paths.DataDir(paths.OSUnixLike)
+)
+
+type app struct {
+	name           string
+	clusterName    string
+	modelUUID      string
+	modelName      string
+	deploymentType caas.DeploymentType
+	client         ecsiface.ECSAPI
+	clock          clock.Clock
+}
+
+func newApplication(
+	name string,
+	clusterName string,
+	modelUUID string,
+	modelName string,
+	deploymentType caas.DeploymentType,
+	client ecsiface.ECSAPI,
+	clock clock.Clock,
+) caas.Application {
+	// TODO: prefix modelName to all resource names?
+	// Because ecs doesnot have namespace!!!
+	// name = modelName + "-" + name
+	return &app{
+		name:           name,
+		clusterName:    clusterName,
+		modelUUID:      modelUUID,
+		modelName:      modelName,
+		deploymentType: deploymentType,
+		client:         client,
+		clock:          clock,
+	}
+}
+
+func (a *app) labels() map[string]*string {
+	// TODO
+	return map[string]*string{
+		"App":       aws.String(a.name),
+		"ModelName": aws.String(a.modelName),
+		"ModelUUID": aws.String(a.modelUUID),
+	}
+}
+
+// Delete deletes the specified application.
+func (a *app) Delete() error {
+	return nil
+}
+
+func strPtrSlice(in ...string) (out []*string) {
+	for _, v := range in {
+		out = append(out, aws.String(v))
+	}
+	return out
+}
+
+func (a *app) volumeName(storageName string) string {
+	return fmt.Sprintf("%s-%s", a.name, storageName)
+}
+
+// getMountPathForFilesystem returns mount path.
+func getMountPathForFilesystem(idx int, appName string, fs jujustorage.KubernetesFilesystemParams) string {
+	if fs.Attachment != nil {
+		return fs.Attachment.Path
+	}
+	return fmt.Sprintf(
+		"%s/fs/%s/%s/%d",
```

Comment:
Are we using host-based volumes here? If so, how are we going to deal with the possibility that two task instances of the same app get scheduled on the same node?

Reply:
We don't use any host-based volumes.
For juju related volumes, we use `DockerVolumeConfiguration` with local driver 

- https://docs.aws.amazon.com/AmazonECS/latest/developerguide/docker-volumes.html.
- https://github.com/juju/juju/pull/12406/files#diff-2ccee37d2965d4d7d708a406d87294f7be801db49dde62eec02686efbbee5709R257

For user-requested volumes, we use `EBS` 
- https://github.com/juju/juju/pull/12406/files#diff-2ccee37d2965d4d7d708a406d87294f7be801db49dde62eec02686efbbee5709R132

Reply:
We don't use any host-based volumes.
For juju related volumes, we use `DockerVolumeConfiguration` with local driver 

- https://docs.aws.amazon.com/AmazonECS/latest/developerguide/docker-volumes.html.
- https://github.com/juju/juju/pull/12406/files#diff-2ccee37d2965d4d7d708a406d87294f7be801db49dde62eec02686efbbee5709R257

For user-requested volumes, we use `EBS` 
- https://github.com/juju/juju/pull/12406/files#diff-2ccee37d2965d4d7d708a406d87294f7be801db49dde62eec02686efbbee5709R132